src/scheduler/Scheduler.cpp

```diff
@@ -191,267 +192,271 @@ std::shared_ptr<InMemoryMessageQueue> Scheduler::getBindQueue()
-    } else {
-        // Increment the number of hops
-        msg.set_hops(msg.hops() + 1);
-
-        // Share with other host (or just log in test mode)
-        if (isTestMode) {
-            logger->debug("TEST MODE - {} not sharing {} with {} ({} hops)",
-                          thisHost,
-                          funcStrWithId,
-                          bestHost,
-                          msg.hops());
-        } else {
-            logger->debug("Host {} sharing {} with {} ({} hops)",
-                          thisHost,
-                          funcStrWithId,
-                          bestHost,
-                          msg.hops());
-
-            FunctionCallClient c(bestHost);
-            c.shareFunctionCall(msg);
-        }
-
-        executedLocally = false;
-    }
-
-    // Add this message to records if necessary
-    if (isTestMode) {
-        recordedMessagesAll.push_back(msg.id());
-        if (executedLocally) {
-            recordedMessagesLocal.push_back(msg.id());
-        } else {
-            recordedMessagesShared.emplace_back(bestHost, msg.id());
-        }
-    }
+    // TODO - more fine-grained locking. This blocks all functions
+    // Lock the whole scheduler to be safe
+    faabric::util::FullLock lock(mx);
 
-    PROF_END(scheduleCall)
-}
+    // Handle forced local execution
+    if (forceLocal) {
+        logger->debug("Executing {} x {} locally", nMessages, funcStr);
 
-long Scheduler::getTotalWarmNodeCount()
-{
-    long totalCount = 0;
-    for (auto p : nodeCountMap) {
-        totalCount += p.second;
-    }
+        for (int i = 0; i < nMessages; i++) {
+            faabric::Message msg = req.messages().at(i);
 
-    return totalCount;
-}
+            funcQueue->enqueue(msg);
+            incrementInFlightCount(msg);
+            addFaaslets(msg);
 
-std::string opinionStr(const SchedulerOpinion& o)
-{
-    switch (o) {
-        case (SchedulerOpinion::MAYBE): {
-            return "MAYBE";
-        }
-        case (SchedulerOpinion::YES): {
-            return "YES";
+            executed.at(i) = thisHost;
         }
-        case (SchedulerOpinion::NO): {
-            return "NO";
-        }
-        default:
-            return "UNKNOWN";
-    }
-}
+    } else {
+        // If we're not the master host, we need to forward the request back to
+        // the master host. This will only happen if a nested batch execution
+        // happens.
+        if (masterHost != thisHost) {
+            logger->debug("Forwarding {} {} back to master {}",
+                          nMessages,
+                          funcStr,
+                          masterHost);
+
+            FunctionCallClient c(masterHost);
+            c.executeFunctions(req);
+        } else {
+            // At this point we know we're the master host, and we've not been
+            // asked to force full local execution.
+
+            // Work out how many we can handle locally
+            int cores = thisHostResources.cores();
```

Comment:
I thought we were not using the term `cores` anymore?

Reply:
Not using the term `nodes` to refer to Faaslets. Cores here really does mean processor cores

Reply:
I was confused